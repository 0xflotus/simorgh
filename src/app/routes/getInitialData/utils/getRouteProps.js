import { matchRoutes } from 'react-router-config';
import pathOr from 'ramda/src/pathOr';
import services from '../../../lib/config/services/loadableConfig';

const guessAmp = url => {
  return url.includes('.amp') || url.includes('/amp');
};

const guessService = url => {
  const foundService = url
    .split('/')
    .filter(Boolean)
    .find(part => Object.keys(services).includes(part));

  return foundService || 'news';
};

const getRouteProps = (routes, url) => {
  const matchedRoutes = matchRoutes(routes, url);

  const amp = pathOr(null, [0, 'match', 'params', 'amp'], matchedRoutes);
  const service = pathOr(
    null,
    [0, 'match', 'params', 'service'],
    matchedRoutes,
  );
  const id = pathOr(null, [0, 'match', 'params', 'id'], matchedRoutes);
  const route = pathOr(null, [0, 'route'], matchedRoutes);
  const match = pathOr(null, [0, 'match'], matchedRoutes);

  return {
    isAmp: amp ? !!amp : guessAmp(url),
    service: service || guessService(url),
    id,
    route,
    match,
  };
};

export default getRouteProps;
