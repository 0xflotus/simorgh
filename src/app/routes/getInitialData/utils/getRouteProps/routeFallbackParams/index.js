import services from '../../../../../lib/config/services/loadableConfig';

export const fallbackAmpParam = url => {
  const ampRegex = /\.amp$/;
  return ampRegex.test(url);
};

export const fallbackServiceParam = url => {
  const pathParts = url.split('/').filter(Boolean);

  if (pathParts[0] && Object.keys(services).includes(pathParts[0])) {
    return pathParts[0];
  }

  return 'news';
};
