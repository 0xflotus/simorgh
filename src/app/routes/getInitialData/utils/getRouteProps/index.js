import { matchRoutes } from 'react-router-config';
import pathOr from 'ramda/src/pathOr';
import { fallbackAmpParam, fallbackServiceParam } from './routeFallbackParams';

const getRouteProps = (routes, url) => {
  const matchedRoutes = matchRoutes(routes, url);

  const amp = pathOr(undefined, [0, 'match', 'params', 'amp'], matchedRoutes);
  const service = pathOr(
    undefined,
    [0, 'match', 'params', 'service'],
    matchedRoutes,
  );
  const id = pathOr(undefined, [0, 'match', 'params', 'id'], matchedRoutes);
  const route = pathOr(undefined, [0, 'route'], matchedRoutes);
  const match = pathOr(undefined, [0, 'match'], matchedRoutes);

  return {
    isAmp: amp ? !!amp : fallbackAmpParam(url),
    service: service || fallbackServiceParam(url),
    id,
    route,
    match,
  };
};

export default getRouteProps;
