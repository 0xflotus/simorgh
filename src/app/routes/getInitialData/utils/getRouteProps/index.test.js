import reactRouterConfig from 'react-router-config';
import getRouteProps from '.';
import * as routeFallbackParams from './routeFallbackParams';

jest.mock('react-router-config');

jest.mock('./routeFallbackParams', () => ({
  fallbackAmpParam: jest.fn().mockImplementation(() => true),
  fallbackServiceParam: jest.fn().mockImplementation(() => 'fallbackService'),
}));

describe('getRouteProps', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('no matched route', () => {
    it('should return an empty object', async () => {
      reactRouterConfig.matchRoutes.mockReturnValue([]);

      try {
        await getRouteProps([], 'fake url');
      } catch (error) {
        expect(error).toMatchSnapshot();
      }

      expect(routeFallbackParams.fallbackAmpParam).not.toHaveBeenCalled();
      expect(routeFallbackParams.fallbackServiceParam).not.toHaveBeenCalled();
    });
  });

  describe('valid route', () => {
    it('should return service, isAmp, route and match', async () => {
      reactRouterConfig.matchRoutes.mockReturnValue([
        {
          route: { route: 'data' },
          match: { params: { service: 'news', amp: undefined } },
        },
      ]);

      const methodCall = await getRouteProps([], 'url');

      expect(routeFallbackParams.fallbackAmpParam).not.toHaveBeenCalled();
      expect(routeFallbackParams.fallbackServiceParam).not.toHaveBeenCalled();

      expect(methodCall).toEqual({
        isAmp: false,
        match: {
          params: {
            amp: undefined,
            service: 'news',
          },
        },
        route: {
          route: 'data',
        },
        service: 'news',
      });
    });
  });

  describe('valid amp route', () => {
    it('should return service, isAmp, route and match', async () => {
      reactRouterConfig.matchRoutes.mockReturnValue([
        {
          route: { route: 'data' },
          match: { params: { service: 'news', amp: '.amp' } },
        },
      ]);

      const methodCall = await getRouteProps([], 'url');

      expect(routeFallbackParams.fallbackAmpParam).not.toHaveBeenCalled();
      expect(routeFallbackParams.fallbackServiceParam).not.toHaveBeenCalled();

      expect(methodCall).toEqual({
        isAmp: true,
        match: {
          params: {
            amp: '.amp',
            service: 'news',
          },
        },
        route: {
          route: 'data',
        },
        service: 'news',
      });
    });
  });

  describe('unknown error route', () => {
    const route = { route: 'data', pageType: 'error' };
    // This is the match returned for a 'catch all' route.
    const match = { path: '/', url: '/', params: {}, isExact: false };

    it('should return service, isAmp, route and match', async () => {
      reactRouterConfig.matchRoutes.mockReturnValue([{ route, match }]);

      const methodCall = await getRouteProps([], 'unknownURL');

      expect(routeFallbackParams.fallbackAmpParam).toHaveBeenCalledWith(
        'unknownURL',
      );
      expect(routeFallbackParams.fallbackServiceParam).toHaveBeenCalledWith(
        'unknownURL',
      );

      expect(methodCall).toEqual({
        isAmp: true,
        match,
        route,
        service: 'fallbackService',
      });
    });
  });
});
