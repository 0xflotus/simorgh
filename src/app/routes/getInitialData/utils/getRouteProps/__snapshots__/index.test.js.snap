// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRouteProps no matched route should return an empty object 1`] = `[TypeError: (0 , _routeFallbackParams.fallbackServiceParam) is not a function]`;
